### ISSUE TITLE
[tune] PBT continues training a sample indefinitely

### ISSUE THREAD

**Issue by HristoBuyukliev** [ray-project/ray#11499]
I want to train a model with PopulationBasedTraining. More specifically, I want to train 8 samples for 40 epochs. However, when I do that, 7 of them are trained for 20-30 epochs, and the last one keeps training indefinitely (it's currently on epoch 340). 

Below I've pasted the relevant piece of code. 

### Reproduction (REQUIRED)
    class PBT(Trainable):
        def setup(self, config):
            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
            self.config = config
            self.best_acc = 0.0
            self.trainer, self.criterion, self.optimizer_conv, self.exp_lr_scheduler = baseline_model(config)
            self.datasets = [FolderDataset(folder, name) for name, folder in config['datasets'].items()]
            self.dataloaders = [DataLoader(dataset, batch_size=16, shuffle=True, num_workers=1) for dataset in self.datasets]
            self.dataset_sizes = [len(dataloader) for dataloader in self.dataloaders]

        def step(self):
            print("training for iteration: {}".format(self.iteration))
            print("parameters for iteration")
            print(self.config)

            losses = train_model(self.trainer, 
                                 self.criterion, 
                                 self.optimizer_conv, 
                                 self.exp_lr_scheduler, 
                                 self.dataloaders, 
                                 num_epochs = 1)        
            auto_garbage_collect()
            return losses

        def save_checkpoint(self, checkpoint_dir):
            checkpoint_path = os.path.join(checkpoint_dir, "model.pth")
            torch.save(self.trainer.state_dict(), checkpoint_path)
            return checkpoint_path

        def load_checkpoint(self, checkpoint_path):
            self.trainer.load_state_dict(torch.load(checkpoint_path))

        def reset_config(self, new_config):
            self.config = new_config
            return True

    def startPBT(perturbation_interval, num_epochs, num_samples, checkpoint_freq, datasets, name):
        torch.manual_seed(12241)
        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

        ray.init(num_cpus=20, num_gpus = 1)

        sched = PopulationBasedTraining(
            time_attr="training_iteration",
            metric="real_validation_acc",
            mode="max",
            perturbation_interval=perturbation_interval,
            hyperparam_mutations={
                "lr": lambda: np.random.lognormal(-4, 1), 
                "weight_decay" : lambda : np.random.lognormal(-4, 1),
            })

        analysis = tune.run(
            PBT,
            name = name,
            scheduler = sched,
            reuse_actors = True,
            stop = {
                "training_iteration": num_epochs
            },
            resources_per_trial={
                "cpu": 20,
                "gpu": 1
            },
            num_samples = num_samples,
            checkpoint_freq = checkpoint_freq,
            config = {
                    "lr" : 0.001, #np.random.lognormal(-6, 2), 
                    "weight_decay" : 0.001, #np.random.lognormal(-5, 1.5), 
                    "local_dir": "/root/",
                    "linear": True,
                    "dropout" : 0.5,
                    "blocks_to_unfreeze": "3_and_4",
                    "hr_message_channels": 256,
                    "lr_scheduler" : "step",
                    "base_lr": 0.001,
                    "max_lr": 0.1,
                    "alpha" : 0.3,
                    "step_size" : 7,
                    "gamma" : 0.18,
                    "momentum" : 0.9,
                    "back_bone" : "resnet18",
                    "batch_size": 32,
                    "datasets": datasets,
                    "name": name
                  },
            verbose = 0,
            local_dir = f"reports/{name}"
            )
        print("Best config is:", analysis.get_best_config(metric="real_validation_acc", mode='max'))
        ray.shutdown()
If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by richardliaw** [comment]
What's the value of `training_iteration: num_epochs`?

Do you have any output/csv of the last trained model too?

**Comment by HristoBuyukliev** [comment]
@richard4912 num_epochs is 40. I deleted the folder, as I needed to clear some space, but the results looked the same as the regular output, only longer. Do you have anything specific you are looking for? I might run the experiment again and check the csv for you. 

I found a workaround, where I set synch=True, as argument to PBT, which enforces models to be trained to the same epoch. Still, I'd very much like to know why this happens.

**Comment by richardliaw** [comment]
@HristoBuyukliev that's good to hear.

Hmm, I'm mainly trying to understand why:
```
        stop = {
            "training_iteration": num_epochs
        },
```
isn't triggering the stopping correctly. I figured there might be something in the output or logs that would help me!

**Comment by Shao-kun-Zhang** [comment]
objective | eval_loss | eval_acc | epoch | time_this_iter_s | should_checkpoint | done | timesteps_total | episodes_total | training_iteration | experiment_id | date | timestamp | time_total_s | pid | hostname | node_ip | time_since_restore | timesteps_since_restore | iterations_since_restore | trial_id
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0.6606498194945850 | 2.127528190612790 | 0.6606498194945850 | 25.0 | 562.5749394893650 | TRUE | FALSE |   |   | 1 | d3b0f5ff86204c21afe01d0917718bac | 2021-04-20_14-46-48 | 1618901208 | 562.5749394893650 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 562.5749394893650 | 0 | 1 | bd73c_00003
0.6425992779783390 | 1.7388126850128200 | 0.6425992779783390 | 100.0 | 1141.7624199390400 | TRUE | FALSE |   |   | 2 | 371c86ae83654b8ba05df1075a2677d4 | 2021-04-20_15-06-04 | 1618902364 | 1709.9914691448200 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 1141.7624199390400 | 0 | 1 | bd73c_00003
0.6750902527075810 | 1.8424313068389900 | 0.6750902527075810 | 51.282051282051300 | 308.0833206176760 | TRUE | FALSE |   |   | 4 | 371c86ae83654b8ba05df1075a2677d4 | 2021-04-20_15-11-29 | 1618902689 | 2026.9642324447600 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 308.0833206176760 | 0 | 1 | bd73c_00003
0.6931407942238270 | 2.429642915725710 | 0.6931407942238270 | 64.1025641025641 | 311.7120985984800 | TRUE | FALSE |   |   | 5 | 371c86ae83654b8ba05df1075a2677d4 | 2021-04-20_15-16-40 | 1618903000 | 2338.6763310432400 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 619.795419216156 | 0 | 2 | bd73c_00003
0.6895306859205780 | 2.555601119995120 | 0.6895306859205780 | 76.92307692307690 | 310.7312910556790 | TRUE | FALSE |   |   | 6 | 371c86ae83654b8ba05df1075a2677d4 | 2021-04-20_15-21-51 | 1618903311 | 2649.4076220989200 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 930.5267102718350 | 0 | 3 | bd73c_00003
0.6823104693140800 | 2.6277050971984900 | 0.6823104693140800 | 89.74358974358970 | 311.2292537689210 | TRUE | FALSE |   |   | 7 | 371c86ae83654b8ba05df1075a2677d4 | 2021-04-20_15-27-02 | 1618903622 | 2960.6368758678400 | 118 | c2a2d5200a1a2011eb0a2730e35ba49cf5f6-zhengxw13-0 | 10.33.116.182 | 1241.7559640407600 | 0 | 4 | bd73c_00003
